### PR TITLE
Simplify the implementation of `run_tv_casting_test.py` for easier addition of future test sequences that verifies that Linux tv-casting-app continues to work with Linux tv-app.

### PR DESCRIPTION
**Pull request by shaoltan-amazon**
**Problem**
There is currently a CI check [implementation](https://github.com/project-chip/connectedhomeip/blob/master/scripts/tests/run_tv_casting_test.py) that verifies that the Linux tv-casting-app continues to work with the Linux tv-app whenever a PR is created. However, adding new test sequences is not that simple as we would need to keep defining new functions for parsing output messages of interest and/or sending input commands. Therefore, we want to improve the design flow to make it easier to add new test cases in the future.

**Solution**
The new design will allow users to define a new test `Sequence` consisting of a test sequence name and a list of `Step` objects. Each `Step` contains the relevant information needed to either parse for output string(s) of interest or send input command. Essentially, to add a new flow, we need to define a test sequence list of `Step` objects containing:
- `subprocess` to parse for output_msg or send input_cmd
- `timeout_sec` specified the timeout duration for parsing the `output_msg` (optional, defaults to DEFAULT_TIMEOUT_SEC)
- `output_msg` or `input_cmd` (mutually exclusive)

Note that there could be multiple instances of a particular output message block so we would need to parse them block-by-block. For this reason, we would need to define the start line, relevant lines we are looking for within the message block, and the last line. If the last line contains trivial closing characters (e.g., closing brackets, braces, or commas), we also need to include the line before it with actual log content. For example:
`Step(subprocess_='tv-casting-app', output_msg=['InvokeResponseMessage =', 'exampleData', 'InteractionModelRevision =', '},'])`

For input commands, we need to define the command as a string followed by `\n`. For example:
`Step(subprocess_='tv-casting-app', input_cmd='cast request 0\n')`

In short, this PR aims to improve the existing test script implementation to follow a new design flow where user can provide test sequence(s) consisting of test steps of output strings or input commands that the test should validate for.

**Testing**
Tested the `run_tv_casting_test.py` script locally to ensure that it still works as expected.

### CODE REVIEW

File: scripts/tests/run_tv_casting_test.py [hunk 1]
```diff
@@ -82,20 +107,79 @@ class LogValueExtractor:
...
+
+    A Step object contains attributes relevant to a test step.
+    """
+
+    # The maximum default time to wait while parsing for output string(s).
+    DEFAULT_TIMEOUT_SEC = 10
+
+    def __init__(
+        self,
+        subprocess_: str,
```
Comment (sharadb-amazon) [inline scripts/tests/run_tv_casting_test.py:157]: subprocess sounds like Python ProcessManager terminology. In the context of tv- and tv-casting-app test sequences, we can give this data member a more meaningful name like `app`

File: scripts/tests/run_tv_casting_test.py [hunk 2]
```diff
@@ -82,20 +107,79 @@ class LogValueExtractor:
...
-                return extract_value_from_string(line, value_name, self.casting_state, self.log_paths)
+                return extract_value_from_string(line, value_name, self.log_paths)
             except ValueError:
                 logging.error(f'Failed to extract `{value_name}` value from line: {line}')
-                handle_casting_failure(self.casting_state, self.log_paths)
+                handle_casting_failure(self.log_paths)
         return None
 
 
+class CommissionerNumberHandler:
```
Comment (sharadb-amazon) [inline scripts/tests/run_tv_casting_test.py:123]: Do we really need to dynamically find the commissioner number? I believe in a controlled test environment, this should be a fixed number (0).

File: scripts/tests/run_tv_casting_test.py [hunk 3]
```diff
@@ -156,453 +224,153 @@ def extract_value_from_string(line: str, value_name: str, casting_state: str, lo
...
-    start_wait_time = time.time()
-
-    while True:
-        # Check if we exceeded the maximum wait time for validating commissioning success between the Linux tv-casting-app and the Linux tv-app.
-        if time.time() - start_wait_time > COMMISSIONING_STAGE_MAX_WAIT_SEC:
-            logging.error(
-                'The commissioning between the Linux tv-casting-app process and the Linux tv-app process did not complete successfully within the timeout.')
-            return False
+def get_general_flow_test_sequence(commissioner_number_handler: CommissionerNumberHandler) -> List[Step]:
+    """Retrieve the test sequence pertaining to the general flow for validating the casting experience between the Linux tv-casting-app and the Linux tv-app."""
```
[the remark this answers is not in the document]
Comment (sharadb-amazon) [reply scripts/tests/run_tv_casting_test.py:236]: Also, is it possible to move these test sequences along with the constants from lines 64-80, to a separate file from the rest of the logic here? If not, can we just move this def to the very top of this file, say at line 81?

File: scripts/tests/run_tv_casting_test.py [hunk 4]
```diff
@@ -82,20 +107,79 @@ class LogValueExtractor:
...
+        self._current_commissioner_number = new_value
+
+    def is_valid(self) -> bool:
+        return self._current_commissioner_number != self.INVALID_COMMISSIONER_NUMBER
+
+
+class Step:
+    """A class to represent a step in a test sequence for validation.
+
+    A Step object contains attributes relevant to a test step.
```
Comment (sharadb-amazon) [inline scripts/tests/run_tv_casting_test.py:149]: Can we represent _**starting**_ and _**stopping**_ the apps using `Steps`?

File: scripts/tests/run_tv_casting_test.py [hunk 5]
```diff
@@ -82,20 +107,79 @@ class LogValueExtractor:
...
+
+    # The maximum default time to wait while parsing for output string(s).
+    DEFAULT_TIMEOUT_SEC = 10
+
+    def __init__(
+        self,
+        subprocess_: str,
+        timeout_sec: Optional[int] = DEFAULT_TIMEOUT_SEC,
+        output_msg: Optional[List[str]] = None,
+        input_cmd: Optional[str] = None,
```
Comment (sharadb-amazon) [inline scripts/tests/run_tv_casting_test.py:160]: Add some inline documentation as well for these data members.

File: scripts/tests/run_tv_casting_test.py [hunk 6]
```diff
@@ -21,25 +21,48 @@
...
 import click
 
+"""
+This test script automates the verification of the casting experience between the Linux tv-casting-app and the Linux tv-app.
+
+It checks for expected output lines from the tv-casting-app and the tv-app in a deterministic order. If these lines are not 
+found, it indicates an issue with the casting experience.
+
+The script currently verifies the general casting flow that includes discovery, commissioning, launchURL, and subscription 
+state. Future updates will enable testing different workflows like the commissioner-generated passcode flow.
```
Comment (sharadb-amazon) [inline scripts/tests/run_tv_casting_test.py:35]: This should just be tracked as a work item. Current status documented in code becomes quickly becomes inaccurate/confusing.

File: scripts/tests/run_tv_casting_test.py [hunk 7]
```diff
@@ -21,25 +21,48 @@
...
+    `Step(subprocess_='tv-casting-app', output_msg=['InvokeResponseMessage =', 'exampleData', 'InteractionModelRevision =', '},'])`
+
+For input commands, define the command string with placeholders for variables that need to be updated. For example:
+    `Step(subprocess_='tv-casting-app', input_cmd='cast request {valid_discovered_commissioner_number}\n', 
+        commissioner_number_handler=commissioner_number_handler)`
+The `commissioner_number_handler` is passed in to help update the `valid_discovered_commissioner_number` placeholder with the 
+actual valid discovered commissioner number.
+
+Users should add a click.option for the desired test flow and update the `examples-linux-tv-casting-app.yaml` workflow file for 
+CI testing of additional flows (to be implemented in future PR(s)).
```
Comment (sharadb-amazon) [inline scripts/tests/run_tv_casting_test.py:56]: Same comment as above.

File: scripts/tests/run_tv_casting_test.py [hunk 8]
```diff
@@ -82,20 +107,79 @@ class LogValueExtractor:
...
+        subprocess_: str,
+        timeout_sec: Optional[int] = DEFAULT_TIMEOUT_SEC,
+        output_msg: Optional[List[str]] = None,
+        input_cmd: Optional[str] = None,
+        commissioner_number_handler: Optional[CommissionerNumberHandler] = None
+    ):
+        self.subprocess_ = subprocess_
+        self.timeout_sec = timeout_sec
+        self.output_msg = output_msg
+        self.input_cmd = input_cmd
```
[the remark this answers is not in the document]
Comment (shaoltan-amazon) [reply scripts/tests/run_tv_casting_test.py]: Good call. This is addressed in the latest commit.

File: scripts/tests/run_tv_casting_test.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (shaoltan-amazon) [reply scripts/tests/run_tv_casting_test.py]: Updated and moved to a separate file `scripts/tests/linux/tv_casting_test_sequences.py`.

File: scripts/tests/run_tv_casting_test.py [hunk 9]
```diff
@@ -156,453 +224,153 @@ def extract_value_from_string(line: str, value_name: str, casting_state: str, lo
...
-        if time.time() - start_wait_time > COMMISSIONING_STAGE_MAX_WAIT_SEC:
-            logging.error(
-                'The commissioning between the Linux tv-casting-app process and the Linux tv-app process did not complete successfully within the timeout.')
-            return False
+def get_general_flow_test_sequence(commissioner_number_handler: CommissionerNumberHandler) -> List[Step]:
+    """Retrieve the test sequence pertaining to the general flow for validating the casting experience between the Linux tv-casting-app and the Linux tv-app."""
 
-        tv_casting_line = tv_casting_app_process.stdout.readline()
-        tv_app_line = tv_app_process.stdout.readline()
+    general_flow_test_sequence = [
```
[the remark this answers is not in the document]
Comment (shaoltan-amazon) [reply scripts/tests/run_tv_casting_test.py]: Great idea! Made the changes in the latest commit.

File: scripts/tests/run_tv_casting_test.py (hunk 1 above)
Comment (shaoltan-amazon) [reply scripts/tests/run_tv_casting_test.py]: Good call. I made the changes in the latest commit.

File: scripts/tests/run_tv_casting_test.py (hunk 2 above)
Comment (shaoltan-amazon) [reply scripts/tests/run_tv_casting_test.py]: Good callout. Made the updates in the latest commit.

File: scripts/tests/run_tv_casting_test.py (hunk 5 above)
Comment (shaoltan-amazon) [reply scripts/tests/run_tv_casting_test.py]: Added them in the latest commit.

File: scripts/tests/run_tv_casting_test.py (hunk 6 above)
Comment (shaoltan-amazon) [reply scripts/tests/run_tv_casting_test.py]: Good point. This has been removed in the latest commit.

File: scripts/tests/run_tv_casting_test.py (hunk 7 above)
Comment (shaoltan-amazon) [reply scripts/tests/run_tv_casting_test.py]: Removed in the latest commit.

File: scripts/tests/run_tv_casting_test.py (hunk 4 above)
Comment (shaoltan-amazon) [reply scripts/tests/run_tv_casting_test.py]: I don't think it would be possible to represent the _**starting**_ the apps using `Steps` since we start the subprocesses with the _ProcessManager_. The _ProcessManager_ also takes care of stopping the subprocesses once we are done with them.

File: scripts/tests/linux/tv_casting_test_sequences.py [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+        self,
+        app: App,
+        timeout_sec: Optional[int] = DEFAULT_TIMEOUT_SEC,
+        output_msg: Optional[List[str]] = None,
+        input_cmd: Optional[str] = None,
+    ):
+        # Validate that either `output_msg` or `input_cmd` is provided, but not both.
+        if output_msg is not None and input_cmd is not None:
+            logging.error('Step cannot contain both `output_msg` and `input_cmd`. Either `output_msg` or `input_cmd` should be provided.')
+            sys.exit(1)
```
Comment (sharadb-amazon) [inline scripts/tests/linux/tv_casting_test_sequences.py:82]: sys.exit is a bit too drastic in constructors. Can we raise an exception in these cases instead of doing a sys.exit here?

File: scripts/tests/run_tv_casting_test.py (hunk 4 above)
Comment (sharadb-amazon) [reply scripts/tests/run_tv_casting_test.py]: Not sure I followed why this isn't possible? Can we have reserved input\_cmd keywords for "Start" and "Stop" that can run the code to call the ProcessManager?

File: scripts/tests/linux/tv_casting_test_sequences.py [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+def get_test_sequence_by_name(test_sequences: List[Sequence], test_sequence_name: str) -> Optional[Sequence]:
+    """Retrieve a test sequence from a list of sequences by its name."""
+
+    for sequence in test_sequences:
+        if sequence.name == test_sequence_name:
+            return sequence
+    return None
+
+
+def get_test_sequences() -> List[Sequence]:
```
Comment (sharadb-amazon) [inline scripts/tests/linux/tv_casting_test_sequences.py:121]: Is this a static method? Can we decorate this with @staticmethod. Same comment for get_test_sequence_by_name

File: scripts/tests/run_tv_casting_test.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sharadb-amazon) [reply scripts/tests/run_tv_casting_test.py]: Looks better now, but why did we also have to move the Step, Sequence and App classes to scripts/tests/linux/tv\_casting\_test\_sequences.py? Is it not possible to just move the test\_sequences array and keep all the logic in the old scripts/tests/run\_tv\_casting\_test.py?

File: scripts/tests/run_tv_casting_test.py (hunk 4 above)
Comment (sharadb-amazon) [reply scripts/tests/run_tv_casting_test.py]: Not sure I followed why this is not possible. Can't we define reserved input\_cmd keywords for "Start" and "Stop" that run the ProcessManager code we need to run?

File: scripts/tests/linux/tv_casting_test_sequences.py (hunk 11 above)
Comment (shaoltan-amazon) [reply scripts/tests/linux/tv_casting_test_sequences.py]: Addressed in the latest commit.

File: scripts/tests/linux/tv_casting_test_sequences.py (hunk 10 above)
Comment (shaoltan-amazon) [reply scripts/tests/linux/tv_casting_test_sequences.py]: Addressed in the latest commit.

File: scripts/tests/run_tv_casting_test.py (hunk 4 above)
Comment (shaoltan-amazon) [reply scripts/tests/run_tv_casting_test.py]: Addressed in the latest commit.

File: scripts/tests/linux/tv_casting_test_sequences.py [hunk 12]
```diff
@@ -14,170 +14,139 @@
...
+for validating the casting experience between the Linux tv-casting-app and the Linux tv-app. 
 
+At the beginning of each test sequence we need to indicate the start up of the tv-app using the `START_APP` string as the `input_cmd`
+followed by the same for the tv-casting-app. On the other hand, at the end of each test sequence we need to ensure that each app will 
+be stopped by providing the `STOP_APP` string as the `input_cmd`. As noted in the example below of `example_test_sequence`, the first
+four steps pertain to starting the apps while the last two are for signaling stopping the apps. 
 
-class App(Enum):
-    """An enumeration of the supported applications."""
+Note: `START_APP` and `STOP_APP` are reserved for signaling the starting and stopping of an apps.
```
Comment (sharadb-amazon) [inline scripts/tests/linux/tv_casting_test_sequences.py:28]: typo: "... stopping of apps."

**Review by sharadb-amazon — APPROVED** [review]

File: scripts/tests/linux/tv_casting_test_sequences.py (hunk 12 above)
Comment (shaoltan-amazon) [reply scripts/tests/linux/tv_casting_test_sequences.py]: Good catch. Fixed in the latest commit.

**Review by chrisdecenzo — APPROVED** [review]

**Review by hawk248 — APPROVED** [review]
